### PR TITLE
Add resources for ACLs & tokens management (new ACLs v1.4)

### PR DESCRIPTION
**Pull request by bhuisgen**
This PR adds the support of these resources:
- `consul_acl_policy` to create new ACL policies
- `consul_acl_token` to create new ACL tokens (management, client and agent tokens)
- `consul_acl_master_token` to setup and bootstrap a cluster only with terraform

To test the ACL policy + token creation with this PR:

````
resource "consul_acl_policy" "test" {
	name = "test"
	rules = "node_prefix \"\" { policy = \"read\" }"
	datacenters = [ "dc1" ]
}

resource "consul_acl_token" "test" {
	description = "test"
	policies = ["${consul_acl_policy.test.name}"]
	local = true
}
````

My goal is to completely create and configure a consul cluster with this resources. Here is the plan to do it (without HTTPS & docker stuff):

````
provider "consul" {
  alias          = "bootstrap"
  address        = "127.0.0.1:8500"
  scheme         = "http"
  datacenter     = "dc1"
}

provider "consul" {
  alias          = "setup"
  address        = "127.0.0.1:8500"
  scheme         = "http"
  datacenter     = "dc1"
  token          = "${consul_acl_master_token.master.token}"
}

provider "consul" {
  address        = "127.0.0.1:8500"
  scheme         = "http"
  datacenter     = "dc1"
  token          = "${consul_acl_token.admin.token}"
}

resource "consul_acl_master_token" "master" {
  provider = "consul.bootstrap"
}

resource "null_resource" "bootstrap" {
  depends_on = ["consul_acl_master_token.master"]

  triggers {
    consul_acl_master_token = "${consul_acl_master_token.master.id}"
  }

  provisioner "local-exec" {
    command = "sleep 30"
  }
}

resource "consul_acl_policy" "admin" {
  depends_on = ["null_resource.bootstrap"]

  provider = "consul.setup"

  name = "admin"
  rules = "node_prefix \"\" { policy = \"write\" } key_prefix \"\" { policy = \"write\" } service_prefix \"\" { policy = \"write\" }"
}


resource "consul_acl_token" "admin" {
  provider = "consul.setup"

  description = "admin token"
  policies = ["${consul_acl_policy.admin.name}"]
}

resource "consul_acl_policy" "agent" {
  depends_on = ["consul_acl_token.admin"]

  provider = "consul.setup"

  name = "agent"
  rules = "node_prefix \"\" { policy = \"write\" } service_prefix \"\" { policy = \"read\" }"
}

resource "consul_acl_token" "agent" {
  provider = "consul.setup"

  policies = ["${consul_acl_policy.agent.name}"]
}
````

### CODE REVIEW

**Comment by bhuisgen** [comment]
For the setup of a whole cluster with terraform, you can use the resource `consul_acl_master_token` to bootstrap the ACL system.

Here is my plan to setup completely a server with terraform by using different `consul` providers with the right token in each step.

````
#
# bootstrap cluster
#

provider "consul" {
  alias      = "bootstrap"
  address    = "127.0.0.1:10000"
  datacenter = "consul"
}

resource "consul_acl_master_token" "master" {
  provider = "consul.bootstrap"
}

output "1.  master token" {
  value     = "${consul_acl_master_token.master.token}"
  sensitive = true
}

#
# setup management token
#

provider "consul" {
  alias      = "setup"
  address    = "127.0.0.1:10000"
  datacenter = "consul"
  token      = "${consul_acl_master_token.master.token}"
}

resource "consul_acl" "admin" {
  provider = "consul.setup"
  name     = "admin"
  type     = "management"
}

output "2. management token" {
  value     = "${consul_acl.admin.token}"
  sensitive = true
}

#
# standard operations with the default provider
#

provider "consul" {
  address    = "127.0.0.1:10000"
  datacenter = "consul"
  token      = "${consul_acl.admin.token}"
}

resource "consul_acl" "agent" {
  name = "agent"
  type = "client"

  rules = <<EOT
node "" { policy = "write" }
service "" { policy = "read" }
EOT
}

output "3. agent token" {
  value     = "${consul_acl.agent.token}"
  sensitive = true
}

resource "consul_acl" "test" {
  name = "test"
  type = "client"

  rules = <<EOT
node "" { policy = "write" }
service "" { policy = "read" }
EOT
}

````
If someone can test that on a real cluster of 3 nodes. It can help :p

**Comment by nsmeef** [comment]
Whats the ETA for this getting released? Would love to get this into a client.

**Comment by pearkes** [comment]
@nsmeef unfortunately I think this would need some work to utilize the [new ACL APIs](https://www.consul.io/docs/upgrade-specific.html#acl-upgrade). See https://github.com/terraform-providers/terraform-provider-consul/issues/46.

We resisted adding ACLs to the Terraform provider specifically because the only ID that could be utilized by Terraform as an identifier (and as a result stored in Terraform state) was the secret token. This was a mistake in the initial ACL API before we had built Terraform and realized how important these low risk identifiers were!

With Consul 1.4.0, we now have "accessor IDs" which are UUID formatted strings that can be used to lookup ACL tokens and stored as relatively low risk data, therefore being appropriate to put in Terraform state.

@bhuisgen I'd be super happy to work with you to get this PR utilizing the new APIs so we can get it merged, but it'd come with the caveat for users that they'd need to be on Consul 1.4.0+.

**Comment by bhuisgen** [comment]
@pearkes I known for the new ACLs in consul 1.4 and I agree that we can now use accessors (like I did in my PR on the terraform vault provider). I will refactor this feature to support it soon.

**Comment by pearkes** [comment]
@bhuisgen sounds good! Let me know how I can help.

**Comment by peimanja** [comment]
Any update on this one?

**Comment by pearkes** [comment]
Thanks for the ping @peimanja and the work here @bhuisgen -- it looks like you got this updated to 1.4 APIs, are you ready for in-depth review?

File: consul/resource_consul_acl_agent_token.go [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+				Type:        schema.TypeBool,
+				Required:    false,
+				Optional:    true,
+				Default:     false,
+				Description: "Flag to set the token local to the current datacenter.",
+			},
+			"token": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "The token.",
```
Comment (remilapeyre) [inline consul/resource_consul_acl_agent_token.go]: You need to set `Sensitive: true` to all sensitive information.

**Comment by remilapeyre** [comment]
Hi @bhuisgen thank you for all this work, I tried running your tests but was not able to do so. I made the following changes to have them running on my computer without depending on Docker:

```
commit d0ea60b028910c46469a7452aae412884f246f06
Author: Rémi Lapeyre <remi.lapeyre@henki.fr>
Date:   Thu Jan 10 14:58:46 2019 +0100

    Make Consul ACL tests working with development server
    
    By using a consul server with ACL enabled and whose default policy is allow, we
    can keep previous tests and run the new ACL tests against the same server.
    
    The only test which does not work with this configuration yet is for
    consul_acl_master_token since bootstrap is already done. Consul documentation
    (https://www.consul.io/docs/commands/acl/acl-bootstrap.html) says:
    
            If all tokens are lost and you need to bootstrap again you can follow the bootstrap reset procedure.
    
    but I couln't find this procedure in the documentation and reading Consul source
    code I'm not sure it's possible to do this here.
    
    Tests can now be ran with:
    
            consul agent -dev -config-file consul_test.hcl &
            export CONSUL_HTTP_ADDR=http://localhost:8500
            make testacc

diff --git a/GNUmakefile b/GNUmakefile
index c3bb3fb..2bcb5a4 100644
--- a/GNUmakefile
+++ b/GNUmakefile
@@ -50,12 +50,6 @@ test-serv: fmtcheck
 	@docker pull "consul:$(CONSUL_VERSION)"
 	docker run --rm -p 127.0.0.1:8500:8500 "consul:$(CONSUL_VERSION)"
 
-test-serv-acl-with-bootstrap: fmtcheck
-	@docker pull "consul:$(CONSUL_VERSION)"
-	docker run --rm -p 127.0.0.1:8500:8500 \
-	    -e 'CONSUL_LOCAL_CONFIG={ "acl": { "enabled": true, "tokens": { "master": "6b0de9ab-6d95-4af8-a965-78ca35a67018" } } }' \
-		"consul:$(CONSUL_VERSION)" agent -server -client=0.0.0.0 -bootstrap-expect=1
-
 test-serv-acl-without-bootstrap: fmtcheck
 	@docker pull "consul:$(CONSUL_VERSION)"
 	docker run --rm -p 127.0.0.1:8500:8500 \
diff --git a/consul/resource_consul_acl_agent_token.go b/consul/resource_consul_acl_agent_token.go
index 9f2122c..6735594 100644
--- a/consul/resource_consul_acl_agent_token.go
+++ b/consul/resource_consul_acl_agent_token.go
@@ -2,10 +2,10 @@ package consul
 
 import (
 	"fmt"
+	"log"
+
 	consulapi "github.com/hashicorp/consul/api"
 	"github.com/hashicorp/terraform/helper/schema"
-
-	"log"
 )
 
 func resourceConsulACLAgentToken() *schema.Resource {
diff --git a/consul/resource_consul_acl_agent_token_test.go b/consul/resource_consul_acl_agent_token_test.go
index e8fc42c..daddfad 100644
--- a/consul/resource_consul_acl_agent_token_test.go
+++ b/consul/resource_consul_acl_agent_token_test.go
@@ -4,13 +4,15 @@ import (
 	"fmt"
 	"testing"
 
-	consulapi "github.com/hashicorp/consul/api"
 	"github.com/hashicorp/terraform/helper/resource"
 	"github.com/hashicorp/terraform/terraform"
 )
 
 func testAccCheckConsulACLAgentTokenDestroy(s *terraform.State) error {
-	client := testAccProvider.Meta().(*consulapi.Client)
+	client, err := getMasterClient()
+	if err != nil {
+		return err
+	}
 
 	for _, rs := range s.RootModule().Resources {
 		if rs.Type != "consul_acl_agent_token" {
@@ -34,7 +36,7 @@ func TestAccConsulACLAgentToken_basic(t *testing.T) {
 		CheckDestroy: testAccCheckConsulACLAgentTokenDestroy,
 		Steps: []resource.TestStep{
 			{
-				Config: testResourceACLAgentTokenConfig_basic(),
+				Config: testResourceACLAgentTokenConfigBasic,
 				Check: resource.ComposeTestCheckFunc(
 					resource.TestCheckResourceAttr("consul_acl_agent_token.test", "description", "test"),
 					resource.TestCheckResourceAttr("consul_acl_agent_token.test", "policies.#", "1"),
@@ -46,8 +48,7 @@ func TestAccConsulACLAgentToken_basic(t *testing.T) {
 	})
 }
 
-func testResourceACLAgentTokenConfig_basic() string {
-	return `
+const testResourceACLAgentTokenConfigBasic = testAccMasterProviderConfiguration + `
 resource "consul_acl_policy" "test" {
 	name = "test"
 	rules = "node_prefix \"\" { policy = \"write\" } service_prefix \"\" { policy = \"read\" }"
@@ -59,4 +60,3 @@ resource "consul_acl_agent_token" "test" {
 	policies = ["${consul_acl_policy.test.name}"]
 	local = true
 }`
-}
diff --git a/consul/resource_consul_acl_policy.go b/consul/resource_consul_acl_policy.go
index e2dec1f..47fef34 100644
--- a/consul/resource_consul_acl_policy.go
+++ b/consul/resource_consul_acl_policy.go
@@ -2,9 +2,10 @@ package consul
 
 import (
 	"fmt"
+	"log"
+
 	consulapi "github.com/hashicorp/consul/api"
 	"github.com/hashicorp/terraform/helper/schema"
-	"log"
 )
 
 func resourceConsulACLPolicy() *schema.Resource {
diff --git a/consul/resource_consul_acl_policy_test.go b/consul/resource_consul_acl_policy_test.go
index ce10c44..dcfa745 100644
--- a/consul/resource_consul_acl_policy_test.go
+++ b/consul/resource_consul_acl_policy_test.go
@@ -4,13 +4,15 @@ import (
 	"fmt"
 	"testing"
 
-	consulapi "github.com/hashicorp/consul/api"
 	"github.com/hashicorp/terraform/helper/resource"
 	"github.com/hashicorp/terraform/terraform"
 )
 
 func testAccCheckConsulACLPolicyDestroy(s *terraform.State) error {
-	client := testAccProvider.Meta().(*consulapi.Client)
+	client, err := getMasterClient()
+	if err != nil {
+		return err
+	}
 
 	for _, rs := range s.RootModule().Resources {
 		if rs.Type != "consul_acl" {
@@ -34,7 +36,7 @@ func TestAccConsulACLPolicy_basic(t *testing.T) {
 		CheckDestroy: testAccCheckConsulACLPolicyDestroy,
 		Steps: []resource.TestStep{
 			{
-				Config: testResourceACLPolicyConfig_basic(),
+				Config: testResourceACLPolicyConfigBasic,
 				Check: resource.ComposeTestCheckFunc(
 					resource.TestCheckResourceAttr("consul_acl_policy.test", "name", "test"),
 					resource.TestCheckResourceAttr("consul_acl_policy.test", "rules", "node_prefix \"\" { policy = \"read\" }"),
@@ -45,11 +47,9 @@ func TestAccConsulACLPolicy_basic(t *testing.T) {
 	})
 }
 
-func testResourceACLPolicyConfig_basic() string {
-	return `
+const testResourceACLPolicyConfigBasic = testAccMasterProviderConfiguration + `
 resource "consul_acl_policy" "test" {
 	name = "test"
 	rules = "node_prefix \"\" { policy = \"read\" }"
 	datacenters = [ "dc1" ]
 }`
-}
diff --git a/consul/resource_consul_acl_token_test.go b/consul/resource_consul_acl_token_test.go
index 72ffe04..b1d6a4a 100644
--- a/consul/resource_consul_acl_token_test.go
+++ b/consul/resource_consul_acl_token_test.go
@@ -4,13 +4,15 @@ import (
 	"fmt"
 	"testing"
 
-	consulapi "github.com/hashicorp/consul/api"
 	"github.com/hashicorp/terraform/helper/resource"
 	"github.com/hashicorp/terraform/terraform"
 )
 
 func testAccCheckConsulACLTokenDestroy(s *terraform.State) error {
-	client := testAccProvider.Meta().(*consulapi.Client)
+	client, err := getMasterClient()
+	if err != nil {
+		return err
+	}
 
 	for _, rs := range s.RootModule().Resources {
 		if rs.Type != "consul_acl" {
@@ -34,7 +36,7 @@ func TestAccConsulACLToken_basic(t *testing.T) {
 		CheckDestroy: testAccCheckConsulACLTokenDestroy,
 		Steps: []resource.TestStep{
 			{
-				Config: testResourceACLTokenConfig_basic(),
+				Config: testResourceACLTokenConfigBasic,
 				Check: resource.ComposeTestCheckFunc(
 					resource.TestCheckResourceAttr("consul_acl_token.test", "description", "test"),
 					resource.TestCheckResourceAttr("consul_acl_token.test", "policies.#", "1"),
@@ -46,8 +48,7 @@ func TestAccConsulACLToken_basic(t *testing.T) {
 	})
 }
 
-func testResourceACLTokenConfig_basic() string {
-	return `
+const testResourceACLTokenConfigBasic = testAccMasterProviderConfiguration + `
 resource "consul_acl_policy" "test" {
 	name = "test"
 	rules = "node \"\" { policy = \"read\" }"
@@ -59,4 +60,3 @@ resource "consul_acl_token" "test" {
 	policies = ["${consul_acl_policy.test.name}"]
 	local = true
 }`
-}
diff --git a/consul/resource_consul_prepared_query_test.go b/consul/resource_consul_prepared_query_test.go
index 6b08ada..5e694e4 100644
--- a/consul/resource_consul_prepared_query_test.go
+++ b/consul/resource_consul_prepared_query_test.go
@@ -64,21 +64,24 @@ func TestAccConsulPreparedQuery_basic(t *testing.T) {
 	})
 }
 
-func checkPreparedQueryExists(s *terraform.State) bool {
+func checkPreparedQueryExists(s *terraform.State, client *consulapi.Client) bool {
 	rn, ok := s.RootModule().Resources["consul_prepared_query.foo"]
 	if !ok {
 		return false
 	}
 	id := rn.Primary.ID
 
-	client := testAccProvider.Meta().(*consulapi.Client).PreparedQuery()
 	opts := &consulapi.QueryOptions{Datacenter: "dc1"}
-	pq, _, err := client.Get(id, opts)
+	pq, _, err := client.PreparedQuery().Get(id, opts)
 	return err == nil && pq != nil
 }
 
 func testAccCheckConsulPreparedQueryDestroy(s *terraform.State) error {
-	if checkPreparedQueryExists(s) {
+	client, err := getMasterClient()
+	if err != nil {
+		return err
+	}
+	if checkPreparedQueryExists(s, client) {
 		return fmt.Errorf("Prepared query 'foo' still exists")
 	}
 	return nil
@@ -86,7 +89,11 @@ func testAccCheckConsulPreparedQueryDestroy(s *terraform.State) error {
 
 func testAccCheckConsulPreparedQueryExists() resource.TestCheckFunc {
 	return func(s *terraform.State) error {
-		if !checkPreparedQueryExists(s) {
+		client, err := getMasterClient()
+		if err != nil {
+			return err
+		}
+		if !checkPreparedQueryExists(s, client) {
 			return fmt.Errorf("Prepared query 'foo' does not exist")
 		}
 		return nil
@@ -110,10 +117,9 @@ func testAccCheckConsulPreparedQueryAttrValue(attr, val string) resource.TestChe
 	}
 }
 
-const testAccConsulPreparedQueryConfig = `
+const testAccConsulPreparedQueryConfig = testAccMasterProviderConfiguration + `
 resource "consul_prepared_query" "foo" {
 	name = "foo"
-	token = "client-token"
 	stored_token = "pq-token"
 	service = "redis"
 	tags = ["prod"]
@@ -136,10 +142,9 @@ resource "consul_prepared_query" "foo" {
 }
 `
 
-const testAccConsulPreparedQueryConfigUpdate1 = `
+const testAccConsulPreparedQueryConfigUpdate1 = testAccMasterProviderConfiguration + `
 resource "consul_prepared_query" "foo" {
 	name = "baz"
-	token = "client-token"
 	stored_token = "pq-token-updated"
 	service = "memcached"
 	tags = ["prod","sup"]
@@ -162,10 +167,9 @@ resource "consul_prepared_query" "foo" {
 }
 `
 
-const testAccConsulPreparedQueryConfigUpdate2 = `
+const testAccConsulPreparedQueryConfigUpdate2 = testAccMasterProviderConfiguration + `
 resource "consul_prepared_query" "foo" {
 	name = "baz"
 	service = "memcached"
-	token = "client-token"
 }
 `
diff --git a/consul/resource_provider_test.go b/consul/resource_provider_test.go
index effde72..a1324aa 100644
--- a/consul/resource_provider_test.go
+++ b/consul/resource_provider_test.go
@@ -1,9 +1,11 @@
 package consul
 
 import (
+	"fmt"
 	"os"
 	"testing"
 
+	consulapi "github.com/hashicorp/consul/api"
 	"github.com/hashicorp/terraform/config"
 	"github.com/hashicorp/terraform/helper/schema"
 	"github.com/hashicorp/terraform/terraform"
@@ -12,11 +14,52 @@ import (
 var testAccProviders map[string]terraform.ResourceProvider
 var testAccProvider *schema.Provider
 
+// The providers and configuration used to test permissions
+var testAccMasterProviders map[string]terraform.ResourceProvider
+var testAccMasterProvider *schema.Provider
+
+const masterToken = "master-token"
+const testAccMasterProviderConfiguration = `
+provider "consul" {
+	token = "` + masterToken + `"
+}`
+
+func getMasterClient() (*consulapi.Client, error) {
+	rp, err := testAccMasterProviderFactory()
+	client := rp.Meta().(*consulapi.Client)
+	return client, err
+}
+
+func testAccMasterProviderFactory() (*schema.Provider, error) {
+	testAccMasterProvider = Provider().(*schema.Provider)
+	raw := map[string]interface{}{
+		"token": masterToken,
+	}
+	rawConfig, err := config.NewRawConfig(raw)
+	if err != nil {
+		return nil, err
+	}
+
+	err = testAccMasterProvider.Configure(terraform.NewResourceConfig(rawConfig))
+	if err != nil {
+		return nil, err
+	}
+	return testAccMasterProvider, nil
+}
+
 func init() {
 	testAccProvider = Provider().(*schema.Provider)
 	testAccProviders = map[string]terraform.ResourceProvider{
 		"consul": testAccProvider,
 	}
+
+	testAccMasterProvider, err := testAccMasterProviderFactory()
+	if err != nil {
+		panic(fmt.Sprintf("err: %s", err))
+	}
+	testAccMasterProviders = map[string]terraform.ResourceProvider{
+		"consul": testAccMasterProvider,
+	}
 }
 
 func TestResourceProvider(t *testing.T) {
diff --git a/consul_test.hcl b/consul_test.hcl
new file mode 100644
index 0000000..f6c9f5a
--- /dev/null
+++ b/consul_test.hcl
@@ -0,0 +1,10 @@
+ui = true
+
+acl {
+  enabled = true
+  default_policy = "allow"
+
+  tokens {
+    master = "master-token"
+  }
+}

```

**Comment by remilapeyre** [comment]
So far it seems to work very nicely, I will do a more thorough review now that I'm able to run the tests.

Can you add documentation for the new resources in `website/docs`?

**Comment by bhuisgen** [comment]
@remilapeyre your patch can't be applied with `git am` so I commited directly.

File: consul/resource_consul_acl_master_token.go [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+	d.Set("policies", policies)
+	d.Set("local", aclToken.Local)
+	d.Set("token", aclToken.SecretID)
+
+	d.SetId(aclToken.AccessorID)
+
+	return resourceConsulACLMasterTokenRead(d, meta)
+}
+
+func resourceConsulACLMasterTokenRead(d *schema.ResourceData, meta interface{}) error {
```
[the remark this answers is not in the document]
Comment (bhuisgen) [reply consul/resource_consul_acl_master_token.go]: Ok I understand, I can copy the same code from the read function of consul_token resource.

File: consul/resource_consul_acl_master_token.go [hunk 3]
```diff
@@ -71,6 +71,23 @@ func resourceConsulACLMasterTokenCreate(d *schema.ResourceData, meta interface{}
...
+	client := meta.(*consulapi.Client)
+
+	id := d.Id()
+	log.Printf("[DEBUG] Reading ACL master token %q", id)
+
+	q := consulapi.QueryOptions{
+		Token: d.Get("token").(string),
+	}
+	_, _, err := client.ACL().TokenRead(id, &q)
+	if err != nil {
```
Comment (remilapeyre) [inline consul/resource_consul_acl_master_token.go]: Could you check for the error specifically like what is being done in https://github.com/terraform-providers/terraform-provider-consul/blob/d843086ae969d04e43dcdbe850ed1dadfe08a968/consul/resource_consul_prepared_query.go#L162-L169 ?

I fear that network issues, for example, would delete the resource unexpectedly.

File: GNUmakefile [hunk 4]
```diff
@@ -48,13 +48,21 @@ test-compile:
...
 	@docker pull "consul:$(CONSUL_VERSION)"
-	docker run --rm -p 127.0.0.1:8500:8500 "consul:$(CONSUL_VERSION)"
+	docker run --rm -p 127.0.0.1:8500:8500 \
+	-e CONSUL_LOCAL_CONFIG=$(shell cat consul_test_default.json | sed "s/\"/\\\"/g") \
+	"consul:$(CONSUL_VERSION)"
+
+test-serv-acl: fmtcheck
+	@docker pull "consul:$(CONSUL_VERSION)"
+	docker run --rm -p 127.0.0.1:8500:8500 \
+	    -e CONSUL_LOCAL_CONFIG=$(shell cat consul_test_acl.json | sed "s/\"/\\\"/g") \
```
[the remark this answers is not in the document]
Comment (bhuisgen) [reply GNUmakefile]: Volume is useless in case of tests.

**Comment by bhuisgen** [comment]
I agree if it's possible to continue to run tests only from the IDE. A different test suite (and file) should be run for the complete bootstrap.

**Comment by remilapeyre** [comment]
Hi @bhuisgen, thanks for your latest changes, I will work on launching a new Consul development server from the tests so we can make sure all cases are covered.

**Comment by p0pr0ck5** [comment]
Any update on this? Would love to have this feature to be able to programmatically manage ACL policies in particular.

**Comment by remilapeyre** [comment]
Hi @p0pr0ck5, thanks for pinging us on this issue, I should be able to add the last missing tests and do a final review in the next days

**Comment by coachin2au** [comment]
Is there an ETA on this merge?

**Comment by Nowaker** [comment]
Looking forward to getting this merged. Super excited!

**Comment by Nowaker** [comment]
@remilapeyre Thank you so much for continuing your work!

**Comment by remilapeyre** [comment]
Hi @bhuisgen, thanks for all the work you did on this!

I removed `consul_acl_agent_token` from the PR, it seems to me to be the exact same resource as `consul_agent_token`. Can you confirm they are both the same?

I also made some small changes to the PR:
 - using the [ACL bootstrap procedure](https://learn.hashicorp.com/consul/advanced/day-1-operations/acl-guide#ensure-the-acl-system-is-configured-properly) to test `consul_master_token` with the same configuration as the rest of the test
 - Removing some redundant attributes
 - Adding some error checking
 - Adding some tests
 - Updating some Makefile targets so testing would continue to work woth Docker
 - Adding a warning in the documentation regarding the master token being saved in the Terraform state when using `consul_master_token`

**Comment by Nowaker** [comment]
@remilapeyre Regarding tokens in plain text, you could think of a Keybase-driven approach. Encrypt secret values with a PGP public key found under a requested account. The same approach as from aws_iam_access_key. https://www.terraform.io/docs/providers/aws/r/iam_access_key.html

**Comment by bhuisgen** [comment]
Thanks @remilapeyre, you can remove the old resource `consul_master_token`.

**Comment by remilapeyre** [comment]
Thanks @bhuisgen! @Nowaker This is one possibility, I don't know whether users will need it though.

**Comment by Nowaker** [comment]
@remilapeyre PGP is mentioned as the first solution in [Terraform Sensitive Data docs](https://www.terraform.io/docs/state/sensitive-data.html), and three different companies I did projects for used it, I guess users anticipate that. It's a very handy solution.

**Comment by remilapeyre** [comment]
Hi, the change in ACLs in Consul 1.4 where tokens where split in two parts, accessor id and secret id was made so the secret id would not be needed to identify a token. Saving the secret id in the terraform state would negate some of the improvements of this change so it's better not make it available through Terraform now.

Even if the secret ID was available through `consul_acl_master_token`, bootstrapping ACLs with this resource would still require the operator to set the `token` attribute of the provider or the `CONSUL_HTTP_TOKEN` environment variable accordingly afterwards.

Given the drawbacks I think it's better to not have the tokens secret IDs accessible for now. We may find a better way to do this in the future and I opened https://github.com/terraform-providers/terraform-provider-consul/issues/95 to track progress on this issue.

**Comment by Nowaker** [comment]
@remilapeyre I don't agree with that. [iam_access_key](https://www.terraform.io/docs/providers/aws/r/iam_access_key.html) is used to generate AWS_ACCESS_KEY_ID (public part) and AWS_SECRET_ACCESS_KEY (secret part). People use it to provision keys and distribute them to their intended users. `consul_acl_token` would be no different. It's up to Terraform user to decide if they want to use a particular resource.

And like I mentioned before, `iam_access_key` offers `pgp_key` option to optionally encrypts the secret key using PGP. This is the ultimate solution for keeping secrets secure.

**Comment by remilapeyre** [comment]
Hi @Nowaker, do you currently have a use case that currently does not work with the Accessor ID?

**Comment by Nowaker** [comment]
@remilapeyre My use case is generating an individual secret ID for each user (or service accounts). Accessor IDs aren't useful in any way.

**Comment by remilapeyre** [comment]
@Nowaker I merged because I wanted to avoid to delay a new release any longer and adding this attribute will be backward compatible so we can still do it in a next release.

I think Consul will integrate more deeply with ecosystems like AWS or Kubernetes in the next release, especially regarding authentication.  Do you use such ecosystem?

Saving the secret ID when creating a new token will negate part of the benefits gained by splitting Accessor IDs and Secret IDs, would a separate datasource solve your issue?

I suggest we continue exploring possible paths for this issue on https://github.com/terraform-providers/terraform-provider-consul/issues/95 which is still open.

**Comment by Nowaker** [comment]
@remilapeyre Data source would be a viable workaround do as long as I can create a new token and immediately get its secret ID in the same Terraform apply run. I actually never used data resources so I don't know if it follow dependency chain.

That said, having the resource output the secret ID is the perfect solution because a future pgp_key would make things work great, and work great with Keybase workflows.

**Comment by ashald** [comment]
@Nowaker FWIW, you can use https://github.com/ashald/terraform-provider-consulacl/blob/master/docs/data_source_consulacl_token.md to do that. :)

**Pull request merged**